executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,348 @@
+package executor
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/memory"
+	"github.com/pingcap/tidb/util/ranger"
+	"golang.org/x/net/context"
+)
+
+//IndexLookUpHashJoin is the hash join executor
+type IndexLookUpHashJoin struct {
+	IndexLookUpJoin
+}
+
+type indexHashJoinOuterWorker struct {
+	outerWorker
+}
+
+type innerHashWorker struct {
+	innerWorker
+	matchPtrBytes [][]byte
+}
+
+func (e *IndexLookUpHashJoin) finishInnerWorker(r interface{}) {
+	if r != nil {
+		e.joinResultCh <- &indexLookUpResult{err: errors.Errorf("%v", r)}
+	}
+	e.workerWg.Done()
+}
+
+func (e *IndexLookUpHashJoin) waitInnerHashWorkersAndCloseResultChan() {
+	e.workerWg.Wait()
+	close(e.joinResultCh)
+}
+
+// Open implements the IndexLookUpHashJoin Executor interface.
+func (e *IndexLookUpHashJoin) Open(ctx context.Context) error {
+	// Be careful, very dirty hack in this line!!!
+	// IndexLookUpJoin need to rebuild executor (the dataReaderBuilder) during
+	// executing. However `executor.Next()` is lazy evaluation when the RecordSet
+	// result is drained.
+	// Lazy evaluation means the saved session context may change during executor's
+	// building and its running.
+	// A specific sequence for example:
+	//
+	// e := buildExecutor()   // txn at build time
+	// recordSet := runStmt(e)
+	// session.CommitTxn()    // txn closed
+	// recordSet.Next()
+	// e.dataReaderBuilder.Build() // txn is used again, which is already closed
+	//
+	// The trick here is `getStartTS` will cache start ts in the dataReaderBuilder,
+	// so even txn is destroyed later, the dataReaderBuilder could still use the
+	// cached start ts to construct DAG.
+	_, err := e.innerCtx.readerBuilder.getStartTS()
+	if err != nil {
+		return err
+	}
+
+	err = e.children[0].Open(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.memTracker = memory.NewTracker(e.id, e.ctx.GetSessionVars().MemQuotaIndexLookupJoin)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+	e.innerPtrBytes = make([][]byte, 0, 8)
+	e.startWorkers(ctx)
+	return nil
+}
+
+func (e *IndexLookUpHashJoin) startWorkers(ctx context.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupJoinConcurrency
+	resultCh := make(chan *lookUpJoinTask, concurrency)
+	e.resultCh = resultCh
+	workerCtx, cancelFunc := context.WithCancel(ctx)
+	e.cancelFunc = cancelFunc
+	innerCh := make(chan *lookUpJoinTask, concurrency)
+	e.workerWg.Add(1)
+	ow := e.newOuterWorker(resultCh, innerCh)
+	go util.WithRecovery(func() {
+		defer close(ow.innerCh)
+		defer e.workerWg.Done()
+		ow.run(workerCtx)
+	}, nil)
+
+	e.joinResultCh = make(chan *indexLookUpResult, concurrency+1)
```

Comment:
Maybe we can set the limit size same as `task.innerResult` to increase goroutine concurrency while keeping memory under control.